### ISSUE TITLE
[RFE] Support building Graal VM with native-image with OpenJDK 11

### ISSUE THREAD

**Issue by jerboaa** [oracle/graal#2196]
**Describe GraalVM and your environment :**
 - GraalVM version: master, 998163db08da1fd64ec25768ddb5fbaeaa40c39a
 - CE
 - Build Time or run time failure: **build time**
 - JDK version: **11.0.6+10**

```
$ java -version
openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment 18.9 (build 11.0.6+10)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.6+10, mixed mode)
```

I've got latest OpenJDK 11.0.6+10 release retrieved from the upstream project[1]. It would be nice if there was a supported way to use a built-from-source JDK to build graal ce with native-image. Currently I'm getting this:

```
Building GraalVM with native-image in /disk/graal/upstream-sources/graal/sdk/mxbuild/linux-amd64/GRAALVM_LIBGRAAL_JAVA11_BNATIVE-IMAGE_BNATIVE-IMAGE-CONFIGURE_LLP_MJDKSL_NI_NIC_NIL_NJU_SJVMCICOMPILER_SNATIVE-IMAGE-AGENT/graalvm-libgraal-java11-20.1.0-dev
build: Checking SubstrateVM requirements for building ...
Could not find any unresolved JVM_* symbols in static JDK libraries
```

Here is my build script:

```
$ OPENJDK="/path/to/openjdk-11.0.6+10"
$ MX="/path/to/mx --java-home=$OPENJDK"
$ rm -rf substratevm/svmbuild
$ pushd substratevm
$ cat > HelloWorld.java <<EOF
public class HelloWorld {
  public static void main(String[] args) {
    System.out.println("Hello World!");
  }
}
EOF
$ JAVAC="$OPENJDK/bin/javac"
$ $JAVAC HelloWorld.java
$ $MX clean
[...]
$ $MX native-image -H:+ReportExceptionStackTraces HelloWorld testme-helloworld
[...]
Building GraalVM with native-image in /somepath/GRAALVM_LIBGRAAL_JAVA11_BNATIVE-IMAGE_BNATIVE-IMAGE-CONFIGURE_LLP_MJDKSL_NI_NIC_NIL_NJU_SJVMCICOMPILER_SNATIVE-IMAGE-AGENT/graalvm-libgraal-java11-20.1.0-dev
build: Checking SubstrateVM requirements for building ...
Could not find any unresolved JVM_* symbols in static JDK libraries
```

One needs to get a JDK 11 from [2], but that's a black box and Graal CE should support a clean build-from-source bootstrap approach.

[1] https://adoptopenjdk.net/upstream.html?variant=openjdk11&jvmVariant=hotspot
[2] https://github.com/graalvm/labs-openjdk-11/releases

**Comment by jerboaa** [comment]
Would there be interest from the graal community to get this fixed? If so I can help to get this done.

**Comment by jerboaa** [comment]
FYI: http://mail.openjdk.java.net/pipermail/jdk-updates-dev/2020-February/002607.html

**Comment by thomaswue** [comment]
The releases you mention are not a "black box", but built from the associated GitHub repository https://github.com/graalvm/labs-openjdk-11. We are interested in anything that makes it easier to work with OpenJDK and GraalVM sources. There are folks like @adinn that have the capability to commit to both projects. I don't know if there are things you could possibly contribute, so I will defer to him for that question.

**Comment by adinn** [comment]
@thomaswue I asked Severin to help out with this task. He is an OpenJDK reviewer for both the man dev tree and the backport trees. He is also a maintainer for the JDK updates project i.e. one of the small group who judge what should or should not be backported to jdk11u. So, he is actually in a better position than me to ensure that any Graal changes committed to the dev tree are back-ported to the jdk11u tree.

I understand your response to the use of the term 'black box' and agree that it is a far from fair description of the open source JVM that you provide. Perhaps a better description of the problem is that your JVM release is a /relatively/ unknown quantity compared to JVM provided by the normal OS software channel and that this is going to cause concern to many system managers. That is not to suggest that there is any issue with your JVM -- I don't mean or want to cast any doubt your technical proficiency or quality control. This is as much about perception as it is about how things stand.

At the very least system managers (or dev-ops staff) will need to take special action before putting your JVM into deployment to verify and agree it's conformity with the reliability and security guarantees that they trust the software from their normal channels to provide. They will then need to track functional and security updates for this software independently. Finally, they will be concerned about whether incompatibilities may arise between your JVM and the underlying OS whenever they need to upgrade the operating system with security and critical functional patches.

Of course they will also need to do those very same tasks wrt the rest of the Graal distribution. However, I am quite sure that many system managers and dev-ops staff will be much happier if they are able to take the JVM out of that equation (sysops staff in particular are notoriously leary about JVMs).

In consequence, we would like to help keep jkd11u releases fully functional as far as supporting Graal is concerned so as to ensure that there is no pushback on deploying Graal. So, this issue was raised to enable discussion of what we will need to do 1) to arrive at that situation and 2) to remain there. So, Severin's offer of help was not just to ask about how to get stuff back-ported that we need right now but to open a channel for further co-ordination between Graal team members on the one hand and both upstream OpenJDK development and the  OpenJDK Updates project on the other hand.

**Comment by jerboaa** [comment]
Thanks for the clarifications @thomaswue and @adinn. So to phrase the goal of this issue differently, we'd like to help keeping the differences between https://github.com/graalvm/labs-openjdk-11/ and OpenJDK 11u as small as possible. I'll propose a plan about this here once I've done some more research.

**Comment by dougxc** [comment]
In the interests of sharing more broadly the input I had for the project of syncing labs-openjdk-11 with OpenJDK 11, here's the email I sent to @adinn @theRealAph and Volker Simonis on Nov 29th 2019:

Hi Volker, Andrew and Andrew,

Following up on our conversations at the GraalVM workshop, I’m providing details of the differences between Labs JDK 11 and OpenJDK 11u. Note that there are 2 versions of Labs JDK 11:

* [Labs OpenJDK 11](https://github.com/graalvm/labs-openjdk-11) - this is the base JDK for building GraalVM CE 11
* Labs OracleJDK 11 - this is the base JDK for building GraalVM EE 11

The differences fall into these 3 categories:

1. Support for Native Image

[JDK-8232748](https://bugs.openjdk.java.net/browse/JDK-8232748): This builds the static libs needed by native image.

[JDK-8233133](https://bugs.openjdk.java.net/browse/JDK-8233133): Needed to defer initialization of classes generated for lambdas until first use of the lambda instead of eagerly as soon as the generated class is loaded.

[JDK-8232571](https://bugs.openjdk.java.net/browse/JDK-8232571): new sun.misc.Signal("INFO") throws IllegalArgumentException since JDK11

2. Support for GraalVM

[JDK-8232118](JDK-8232118): This is required to produce a GraalVM binary where JVMCI is non-experimental and enabled by default.

[JDK-8232080](https://bugs.openjdk.java.net/browse/JDK-8232080): This allows for JDK images produced by jlink to change the defaults for HotSpot -XX options. This is [used](https://github.com/oracle/graal/blob/88de518340d92d04087eb4075e361451dbbe9d2b/sdk/mx.sdk/mx_sdk_vm.py#L590) to set the EnableJVMCIProduct option introduced by JDK-8232118. It also allows for certain vendor information to be modified in the jlinked image which we use to brand the GraalVM in the output of `java -version`.

3. Support for libgraal

We managed to get all the API changes necessary for libgraal into JDK 11 ([JDK-8204855](https://bugs.openjdk.java.net/browse/JDK-8204855)) which meant we could continue to compile Graal against JDK 8 and JDK 11. We continued to develop the implementation of the JVMCI API in JDK 8 until libgraal worked. Later, these implementation changes were merged to JDK 13 ([JDK-8220623](https://bugs.openjdk.java.net/browse/JDK-8220623)) and from then on further JVMCI changes were pushed in JDK tip. When it came time to support libgraal in JDK 11, Tom made a [large monolithic commit](https://github.com/graalvm/labs-openjdk-11/commit/fbbdf5f15300716110d2ac9b9bd7f70b03b35a7d) to Labs JDK 11 to back port all these libgraal changes and others that we wanted. From that point on, further JVMCI have been made to Labs OpenJDK 11 to fix libgraal bugs and address various other issues.

I’m currently working on integrating the category 1 and 2 commits to Labs OpenJDK 11 (they are already in Labs OracleJDK 11). Once they are merged, it should simply be a matter of diffing Labs OpenJDK 11 against OpenJDK 11u to see what delta remains. A current snapshot of this can be seen at https://github.com/dougxc/labs-openjdk-11/commit/88d482d01203f27728046da18c64049e6bef3c26

-Doug

**Comment by RealCLanger** [comment]
Thanks, Doug, for the info. I'll also have a look into getting this stuff in the OpenJDK 11 repo and work together with Severin.

**Comment by jerboaa** [comment]
As far as native image-capable SVM/Graal build with OpenJDK 11 is concerned [JDK-8232748](https://bugs.openjdk.java.net/browse/JDK-8232748) is the missing piece.

- [x] JDK-8233133: Introduce a system property to disable eager lambda initialization. **Included in OpenJDK 11.0.7+1 (currently EA)**
- [x] JDK-8232571: new sun.misc.Signal("INFO") throws IllegalArgumentException since JDK11. **Included in OpenJDK 11.0.7+1 (currently EA)**
- [ ] JDK-8232748: Build static versions of certain JDK libraries. Currently in [RFR](https://mail.openjdk.java.net/pipermail/jdk-updates-dev/2020-February/002607.html). Looks like it's now waiting on [JDK-8189861](https://bugs.openjdk.java.net/browse/JDK-8189861) for `FindFiles`.

For graal it looks like JDK-8232118 has some dependencies. Like JDK-8220623: [JVMCI] Update JVMCI to support JVMCI based Compiler compiled into shared library. JDK-8232080 is something I've briefly looked at, but it doesn't apply cleanly. I haven't managed to dig further down the rabbit hole yet.

**Comment by RealCLanger** [comment]
I think, key will be to get the "large monolithic" into OpenJDK 11. I'll try this one in jdk11u-dev. Goal should be to have it for 11.0.8. I assume now in rampdown phase of 11.0.7 it'll be too intrusive.

**Comment by jerboaa** [comment]
> I think, key will be to get the "large monolithic" into OpenJDK 11. I'll try this one in jdk11u-dev. Goal should be to have it for 11.0.8.

Cool.

> I assume now in rampdown phase of 11.0.7 it'll be too intrusive.

Yes.

**Comment by tkrodriguez** [comment]
[JDK-8205078:](https://bugs.openjdk.java.net/browse/JDK-8205078) [Graal] org.graalvm.compiler.core.test.VerifyDebugUsageTest fails with "Expected exception: VerificationError"
[JDK-8209535:](https://bugs.openjdk.java.net/browse/JDK-8209535) [JVMCI] Do not swallow NoClassDefFoundError when converting JVMCI methods and fields to reflection objects
[JDK-8206992:](https://bugs.openjdk.java.net/browse/JDK-8206992) Update Graal
[JDK-8209624:](https://bugs.openjdk.java.net/browse/JDK-8209624) [JVMCI] Invalidate nmethods instead of directly unloading them when the InstalledCode is dropped
[JDK-8210066:](https://bugs.openjdk.java.net/browse/JDK-8210066) [JVMCI] iterateFrames uses wrong GrowableArray API for appending
[JDK-8210478:](https://bugs.openjdk.java.net/browse/JDK-8210478) Update Graal
[JDK-8199885:](https://bugs.openjdk.java.net/browse/JDK-8199885) [Graal] org.graalvm.compiler.core.test.CountedLoopTest fails with "ControlFlowAnchor should never be cloned in the same graph"
[JDK-8209136:](https://bugs.openjdk.java.net/browse/JDK-8209136) [JVMCI] Make sure volatile fields are read as volatile during constant reflection.
[JDK-8208686:](https://bugs.openjdk.java.net/browse/JDK-8208686) [AOT] JVMTI ResourceExhausted event repeated for same allocation
[JDK-8207146:](https://bugs.openjdk.java.net/browse/JDK-8207146) Rename jdk.internal.misc.Unsafe::xxxObject to xxxReference
[JDK-8212817:](https://bugs.openjdk.java.net/browse/JDK-8212817) [JVMCI] ResolvedJavaMethod.isInVirtualMethodTable throws InternalError
[JDK-8212934:](https://bugs.openjdk.java.net/browse/JDK-8212934) [JVMCI] do not propagate resolution error in HotSpotResolvedJavaFieldImpl.getType
[JDK-8213203:](https://bugs.openjdk.java.net/browse/JDK-8213203) [JVMCI] adopt formatting changes from jvmci 8
[JDK-8213347:](https://bugs.openjdk.java.net/browse/JDK-8213347) [JVMCI] remove use of reflection in JVMCI
[JDK-8213701:](https://bugs.openjdk.java.net/browse/JDK-8213701) Add org.graalvm.compiler.debug.test.TimerKeyTest to problem list
[JDK-8210777:](https://bugs.openjdk.java.net/browse/JDK-8210777) Update Graal
[JDK-8213907:](https://bugs.openjdk.java.net/browse/JDK-8213907) [JVMCI] avoid Class.getDeclared* methods when converting JVMCI objects to reflection objects
[JDK-8213259:](https://bugs.openjdk.java.net/browse/JDK-8213259) [AOT] AOTing java.base fails with "java.lang.AssertionError: no fingerprint for Ljdk/internal/event/Event"
[JDK-8214023:](https://bugs.openjdk.java.net/browse/JDK-8214023) Update Graal
[JDK-8215224:](https://bugs.openjdk.java.net/browse/JDK-8215224) Update Graal
[JDK-8217445:](https://bugs.openjdk.java.net/browse/JDK-8217445) [JVMCI] incorrect management of JVMCI compilation failure reason string
[JDK-8218074:](https://bugs.openjdk.java.net/browse/JDK-8218074) Update Graal
[JDK-8220389:](https://bugs.openjdk.java.net/browse/JDK-8220389) Update Graal
[JDK-8221341:](https://bugs.openjdk.java.net/browse/JDK-8221341) Update Graal
[JDK-8221598:](https://bugs.openjdk.java.net/browse/JDK-8221598) Update Graal
[JDK-8219403:](https://bugs.openjdk.java.net/browse/JDK-8219403) JVMCIRuntime::adjust_comp_level should be replaced
[JDK-8220623:](https://bugs.openjdk.java.net/browse/JDK-8220623) [JVMCI] Update JVMCI to support JVMCI based Compiler compiled into shared library
[JDK-8222665:](https://bugs.openjdk.java.net/browse/JDK-8222665) Update Graal
[JDK-8223332:](https://bugs.openjdk.java.net/browse/JDK-8223332) Update JVMCI
[JDK-8218700:](https://bugs.openjdk.java.net/browse/JDK-8218700) infinite loop in HotSpotJVMCIMetaAccessContext.fromClass after OutOfMemoryError
[JDK-8223639:](https://bugs.openjdk.java.net/browse/JDK-8223639) [JVMCI] jvmciCompiler.cpp needs to include "oops/objArrayOop.inline.hpp""
[JDK-8223346:](https://bugs.openjdk.java.net/browse/JDK-8223346) Update Graal
[JDK-8223320:](https://bugs.openjdk.java.net/browse/JDK-8223320) [AOT] jck test api/javax_script/ScriptEngine/PutGet.html fails when test classes are AOTed
[JDK-8225019:](https://bugs.openjdk.java.net/browse/JDK-8225019) Update JVMCI
[JDK-8209626:](https://bugs.openjdk.java.net/browse/JDK-8209626) [JVMCI] Use implicit exception table for dispatch and printing
[JDK-8223050:](https://bugs.openjdk.java.net/browse/JDK-8223050) JVMCI: findUniqueConcreteMethod() should not use Dependencies::find_unique_concrete_method() for non-virtual methods
[JDK-8225492:](https://bugs.openjdk.java.net/browse/JDK-8225492) Update JVMCI
[JDK-8223796:](https://bugs.openjdk.java.net/browse/JDK-8223796) JVMCIEnv::get_jvmci_type does not keep klasses alive
[JDK-8226222:](https://bugs.openjdk.java.net/browse/JDK-8226222) [JVMCI] Export AArch64 field VM_Version::_psr_info.dczid_el0
[JDK-8225810:](https://bugs.openjdk.java.net/browse/JDK-8225810) Update JVMCI
[JDK-8225497:](https://bugs.openjdk.java.net/browse/JDK-8225497) Update Graal
[JDK-8226533:](https://bugs.openjdk.java.net/browse/JDK-8226533) JVMCI: findUniqueConcreteMethod should handle statically bindable methods directly
[JDK-8226778:](https://bugs.openjdk.java.net/browse/JDK-8226778) [JVMCI] Handle unpacking properly in Deoptimiziation::get_cached_box()
[JDK-8226566:](https://bugs.openjdk.java.net/browse/JDK-8226566) [JVMCI] java.* classes are no longer necessarily resolved by the boot class loader
[JDK-8228340:](https://bugs.openjdk.java.net/browse/JDK-8228340) JVMCI deleteGlobalHandle doesn't delete the handle
[JDK-8226771:](https://bugs.openjdk.java.net/browse/JDK-8226771) Update Graal
[JDK-8229797:](https://bugs.openjdk.java.net/browse/JDK-8229797) [JVMCI] Clean up no longer used JVMCI::dependencies_invalid value
[JDK-8229201:](https://bugs.openjdk.java.net/browse/JDK-8229201) Update Graal
[JDK-8231973:](https://bugs.openjdk.java.net/browse/JDK-8231973) Update Graal
[JDK-8231659:](https://bugs.openjdk.java.net/browse/JDK-8231659) [JVMCI] Could not work PrintAssembly for JVMCI installed code
[JDK-8231754:](https://bugs.openjdk.java.net/browse/JDK-8231754) [JVMCI] Make r27 unconditionally reserved in JVMCI
[JDK-8232904:](https://bugs.openjdk.java.net/browse/JDK-8232904) Update JVMCI
[JDK-8232118:](https://bugs.openjdk.java.net/browse/JDK-8232118) Add JVM option to enable JVMCI compilers in product mode
[JDK-8233035:](https://bugs.openjdk.java.net/browse/JDK-8233035) Update JVMCI
[JDK-8233273:](https://bugs.openjdk.java.net/browse/JDK-8233273) Update Graal
[JDK-8233745:](https://bugs.openjdk.java.net/browse/JDK-8233745) [JVMCI] TranslatedException should serialize classloader and module info
[JDK-8233900:](https://bugs.openjdk.java.net/browse/JDK-8233900) [JVMCI] improve help text for EnableJVMCIProduct option
[JDK-8233841:](https://bugs.openjdk.java.net/browse/JDK-8233841) Update Graal
[JDK-8234359:](https://bugs.openjdk.java.net/browse/JDK-8234359) [JVMCI] invalidate_nmethod_mirror shouldn't use a phantom reference
[JDK-8234681:](https://bugs.openjdk.java.net/browse/JDK-8234681) Remove UseJVMCIClassLoader logic from JVMCI code

**Comment by RealCLanger** [comment]
> Earlier in this discussion I'd attempted to create a full list of bugs that comprise everything that needs to backported. I didn't produce a complete list but here's the partial list I'd gotten in case it's useful.
> 
> [JDK-8205078:](https://bugs.openjdk.java.net/browse/JDK-8205078) [Graal] org.graalvm.compiler.core.test.VerifyDebugUsageTest fails with "Expected exception: VerificationError"
> [JDK-8209535:](https://bugs.openjdk.java.net/browse/JDK-8209535) [JVMCI] Do not swallow NoClassDefFoundError when converting JVMCI methods and fields to reflection objects
> [JDK-8206992:](https://bugs.openjdk.java.net/browse/JDK-8206992) Update Graal
> [JDK-8209624:](https://bugs.openjdk.java.net/browse/JDK-8209624) [JVMCI] Invalidate nmethods instead of directly unloading them when the InstalledCode is dropped
> [JDK-8210066:](https://bugs.openjdk.java.net/browse/JDK-8210066) [JVMCI] iterateFrames uses wrong GrowableArray API for appending
> [JDK-8210478:](https://bugs.openjdk.java.net/browse/JDK-8210478) Update Graal
> [JDK-8199885:](https://bugs.openjdk.java.net/browse/JDK-8199885) [Graal] org.graalvm.compiler.core.test.CountedLoopTest fails with "ControlFlowAnchor should never be cloned in the same graph"
> [JDK-8209136:](https://bugs.openjdk.java.net/browse/JDK-8209136) [JVMCI] Make sure volatile fields are read as volatile during constant reflection.
> [JDK-8208686:](https://bugs.openjdk.java.net/browse/JDK-8208686) [AOT] JVMTI ResourceExhausted event repeated for same allocation
> [JDK-8207146:](https://bugs.openjdk.java.net/browse/JDK-8207146) Rename jdk.internal.misc.Unsafe::xxxObject to xxxReference
> [JDK-8212817:](https://bugs.openjdk.java.net/browse/JDK-8212817) [JVMCI] ResolvedJavaMethod.isInVirtualMethodTable throws InternalError
> [JDK-8212934:](https://bugs.openjdk.java.net/browse/JDK-8212934) [JVMCI] do not propagate resolution error in HotSpotResolvedJavaFieldImpl.getType
> [JDK-8213203:](https://bugs.openjdk.java.net/browse/JDK-8213203) [JVMCI] adopt formatting changes from jvmci 8
> [JDK-8213347:](https://bugs.openjdk.java.net/browse/JDK-8213347) [JVMCI] remove use of reflection in JVMCI
> [JDK-8213701:](https://bugs.openjdk.java.net/browse/JDK-8213701) Add org.graalvm.compiler.debug.test.TimerKeyTest to problem list
> [JDK-8210777:](https://bugs.openjdk.java.net/browse/JDK-8210777) Update Graal
> [JDK-8213907:](https://bugs.openjdk.java.net/browse/JDK-8213907) [JVMCI] avoid Class.getDeclared* methods when converting JVMCI objects to reflection objects
> [JDK-8213259:](https://bugs.openjdk.java.net/browse/JDK-8213259) [AOT] AOTing java.base fails with "java.lang.AssertionError: no fingerprint for Ljdk/internal/event/Event"
> [JDK-8214023:](https://bugs.openjdk.java.net/browse/JDK-8214023) Update Graal
> [JDK-8215224:](https://bugs.openjdk.java.net/browse/JDK-8215224) Update Graal
> [JDK-8217445:](https://bugs.openjdk.java.net/browse/JDK-8217445) [JVMCI] incorrect management of JVMCI compilation failure reason string
> [JDK-8218074:](https://bugs.openjdk.java.net/browse/JDK-8218074) Update Graal
> [JDK-8220389:](https://bugs.openjdk.java.net/browse/JDK-8220389) Update Graal
> [JDK-8221341:](https://bugs.openjdk.java.net/browse/JDK-8221341) Update Graal
> [JDK-8221598:](https://bugs.openjdk.java.net/browse/JDK-8221598) Update Graal
> [JDK-8219403:](https://bugs.openjdk.java.net/browse/JDK-8219403) JVMCIRuntime::adjust_comp_level should be replaced
> [JDK-8220623:](https://bugs.openjdk.java.net/browse/JDK-8220623) [JVMCI] Update JVMCI to support JVMCI based Compiler compiled into shared library
> [JDK-8222665:](https://bugs.openjdk.java.net/browse/JDK-8222665) Update Graal
> [JDK-8223332:](https://bugs.openjdk.java.net/browse/JDK-8223332) Update JVMCI
> [JDK-8218700:](https://bugs.openjdk.java.net/browse/JDK-8218700) infinite loop in HotSpotJVMCIMetaAccessContext.fromClass after OutOfMemoryError
> [JDK-8223639:](https://bugs.openjdk.java.net/browse/JDK-8223639) [JVMCI] jvmciCompiler.cpp needs to include "oops/objArrayOop.inline.hpp""
> [JDK-8223346:](https://bugs.openjdk.java.net/browse/JDK-8223346) Update Graal
> [JDK-8223320:](https://bugs.openjdk.java.net/browse/JDK-8223320) [AOT] jck test api/javax_script/ScriptEngine/PutGet.html fails when test classes are AOTed
> [JDK-8225019:](https://bugs.openjdk.java.net/browse/JDK-8225019) Update JVMCI
> [JDK-8209626:](https://bugs.openjdk.java.net/browse/JDK-8209626) [JVMCI] Use implicit exception table for dispatch and printing
> [JDK-8223050:](https://bugs.openjdk.java.net/browse/JDK-8223050) JVMCI: findUniqueConcreteMethod() should not use Dependencies::find_unique_concrete_method() for non-virtual methods
> [JDK-8225492:](https://bugs.openjdk.java.net/browse/JDK-8225492) Update JVMCI
> [JDK-8223796:](https://bugs.openjdk.java.net/browse/JDK-8223796) JVMCIEnv::get_jvmci_type does not keep klasses alive
> [JDK-8226222:](https://bugs.openjdk.java.net/browse/JDK-8226222) [JVMCI] Export AArch64 field VM_Version::_psr_info.dczid_el0
> [JDK-8225810:](https://bugs.openjdk.java.net/browse/JDK-8225810) Update JVMCI
> [JDK-8225497:](https://bugs.openjdk.java.net/browse/JDK-8225497) Update Graal
> [JDK-8226533:](https://bugs.openjdk.java.net/browse/JDK-8226533) JVMCI: findUniqueConcreteMethod should handle statically bindable methods directly
> [JDK-8226778:](https://bugs.openjdk.java.net/browse/JDK-8226778) [JVMCI] Handle unpacking properly in Deoptimiziation::get_cached_box()
> [JDK-8226566:](https://bugs.openjdk.java.net/browse/JDK-8226566) [JVMCI] java.* classes are no longer necessarily resolved by the boot class loader
> [JDK-8228340:](https://bugs.openjdk.java.net/browse/JDK-8228340) JVMCI deleteGlobalHandle doesn't delete the handle
> [JDK-8226771:](https://bugs.openjdk.java.net/browse/JDK-8226771) Update Graal
> [JDK-8229797:](https://bugs.openjdk.java.net/browse/JDK-8229797) [JVMCI] Clean up no longer used JVMCI::dependencies_invalid value
> [JDK-8229201:](https://bugs.openjdk.java.net/browse/JDK-8229201) Update Graal
> [JDK-8231973:](https://bugs.openjdk.java.net/browse/JDK-8231973) Update Graal
> [JDK-8231659:](https://bugs.openjdk.java.net/browse/JDK-8231659) [JVMCI] Could not work PrintAssembly for JVMCI installed code
> [JDK-8231754:](https://bugs.openjdk.java.net/browse/JDK-8231754) [JVMCI] Make r27 unconditionally reserved in JVMCI
> [JDK-8232904:](https://bugs.openjdk.java.net/browse/JDK-8232904) Update JVMCI
> [JDK-8232118:](https://bugs.openjdk.java.net/browse/JDK-8232118) Add JVM option to enable JVMCI compilers in product mode
> [JDK-8233035:](https://bugs.openjdk.java.net/browse/JDK-8233035) Update JVMCI
> [JDK-8233273:](https://bugs.openjdk.java.net/browse/JDK-8233273) Update Graal
> [JDK-8233745:](https://bugs.openjdk.java.net/browse/JDK-8233745) [JVMCI] TranslatedException should serialize classloader and module info
> [JDK-8233900:](https://bugs.openjdk.java.net/browse/JDK-8233900) [JVMCI] improve help text for EnableJVMCIProduct option
> [JDK-8233841:](https://bugs.openjdk.java.net/browse/JDK-8233841) Update Graal
> [JDK-8234359:](https://bugs.openjdk.java.net/browse/JDK-8234359) [JVMCI] invalidate_nmethod_mirror shouldn't use a phantom reference
> [JDK-8234681:](https://bugs.openjdk.java.net/browse/JDK-8234681) Remove UseJVMCIClassLoader logic from JVMCI code

The question is whether we should try to go bug by bug through this list or whether it's better to use the monolitic change at once.

**Comment by RealCLanger** [comment]
I played around with the JDK13 monolith (https://github.com/graalvm/labs-openjdk-11/commit/fbbdf5f15300716110d2ac9b9bd7f70b03b35a7d). It actually applies quite well to jdk11u-dev with a few resolves and parts from JDK-8208686. I'll try to run this through our OpenJDK test queue now and will then start a discussion on jdk-updates-dev about merging this in.

Maybe we need to compile a list of all jdk/jdk backports that would be resolved by this big bang.

**Comment by jerboaa** [comment]
We've recently discovered that this commit might be problematic for OpenJDK 11u compatibility. See the comments on the commit:

https://github.com/oracle/graal/commit/8fb8fc502411b3c84308ff5cda4d6c1f775f3920

It relies on this from labs-openjdk-11:
https://github.com/graalvm/labs-openjdk-11/commit/9eee9ee40aeb69c6c85d1492a3369c232807f3c0

We could create a separate issue to track that particular item if that's preferred.

**Comment by mirabilos** [comment]
Oh cool, does this mean I’ll be able to use the native-image AOT compiler with stock OpenJDK 11.0.8 in Debian once that comes out, instead of having to install a second JRE/JDK (with all the system-wide trouble *that* creates, and which I’m therefore not going to do)?

That would mean I should really look into using this…

**Comment by jerboaa** [comment]
@mirabilos This ticket tracks buildability of substrate VM with stock OpenJDK 11.

If you can convince Debian packagers to create a `static-libs` sub-package for OpenJDK 11 similar to what we did for [Fedora](https://src.fedoraproject.org/rpms/java-11-openjdk/pull-request/84) then you *should* be able to build substrate (including native-image) with plain OpenJDK 11.0.8 as build JDK which is a first step towards that goal of yours.

**Comment by mirabilos** [comment]
On Thu, 9 Jul 2020, Dongjin Lee wrote:

> Hello Debian Java community,
>
> It seems like there was a discussion about providing GraalVM community
> edition (aka GraalVM CE) Debian package years ago, but it is still not

Yes, please!

I also asked around upstream, and, according to them, there’s still a
little that can be done in our OpenJDK packaging:


On Tue, 02 Jun 2020 00:45:30 -0700, Severin Gehwolf wrote:

>@mirabilos This ticket tracks buildability of substrate VM with stock OpenJDK 11.
>
>If you can convince Debian packagers to create a `static-libs`
>sub-package for OpenJDK 11 similar to what we did for
>[Fedora](https://src.fedoraproject.org/rpms/java-11-openjdk/pull-request/84)
>then you *should* be able to build substrate (including native-image)
>with plain OpenJDK 11.0.8 as build JDK which is a first step towards
>that goal of yours.

>Reply to this email directly or view it on GitHub:
>https://github.com/oracle/graal/issues/2196#issuecomment-637355855

Perhaps this (also looking at the mentioned issue, which shows a
number of other things) will enable the Free part of GraalVM to
be completely available in Debian.

bye,
//mirabilos
-- 
tarent solutions GmbH
Rochusstraße 2-4, D-53123 Bonn • http://www.tarent.de/
Tel: +49 228 54881-393 • Fax: +49 228 54881-235
HRB 5168 (AG Bonn) • USt-ID (VAT): DE122264941
Geschäftsführer: Dr. Stefan Barth, Kai Ebenrett, Boris Esser, Alexander Steeg

**Comment by jerboaa** [comment]
I should mention here that for latest Graal master tree OpenJDK 11.0.9+ will be required due to:
https://github.com/oracle/graal/issues/2628
Fixed with:
https://bugs.openjdk.java.net/browse/JDK-8249047 (only in the jdk11u-dev tree for now)

**Comment by jerboaa** [comment]
For the record, we build the downstream [Mandrel releases](https://github.com/graalvm/mandrel/releases/) with stock OpenJDK 11u by using a small build script called [mandrel-packaging](https://github.com/graalvm/mandrel-packaging). So as far as we are concerned this issue is resolved. Of course it'll need regular patches every once in a while, but we keep track of breakages within Mandrel.

Note that we didn't bring in all JVMCI changes and libgraal changes into OpenJDK 11u as we've considered them too risky. We keep track of them via the [jdk11u-jvmci-defer](https://bugs.openjdk.java.net/issues/?jql=labels%20%3D%20jdk11u-jvmci-defer) label in JBS.
